pkg/storage/cloud/external_storage.go

```diff
@@ -518,39 +556,74 @@ func checkHTTPContentRangeHeader(h string, pos int64) error {
 	return nil
 }
 
+func (r *resumingHTTPReader) sendRequest(
+	reqHeaders map[string]string,
+) (resp *http.Response, err error) {
+	prev := timeutil.Now()
+	attempt := 0
+	for retries := retry.StartWithCtx(r.ctx, httpRetryOptions); retries.Next(); {
+		attempt++
+		resp, err = r.client.req(r.ctx, "GET", r.url, nil, reqHeaders)
+		if resp != nil {
+			r.body = resp.Body
```

Comment:
nit: i find it a little confusing to have this method mutating `r.body` but then also returning `resp` for callers to read (and they then use it to mutate other fields of `r`), plus sometimes callers also assign `r.body`.

I think it'd be cleaner to either 

a.  not return `resp` at all: this method always handles closing/nil-ing `r.body` first, sets `r. body`, `r.canResume`, and checks the range, or
b. not modify `r` at all: perhaps even making this a function that takes a ctx, client, etc as args, and make callers handle mutating r.